vsintegration/src/vs/FsPkgs/FSharp.Project/FS/Project.fs

```diff
@@ -1464,6 +1464,22 @@ See also ...\SetupAuthoring\FSharp\Registry\FSProjSys_Registration.wxs, e.g.
                 if projectSite.State = ProjectSiteOptionLifetimeState.Opening then
                     // This is the first time, so set up interface for language service to talk to us
                     projectSite.Open(x.CreateRunningProjectSite())
+
+                // refresh which refs (if any) contain type providers
+
+                let timingFile = Environment.ExpandEnvironmentVariables("%userprofile%\\desktop\\TPRefCheckTiming.txt")
+                if File.Exists(timingFile) then
+                    let sw = Stopwatch.StartNew()
+                    for ref in this.GetReferenceContainer().EnumReferences() do
+                        match ref with
+                        | :? TypeProviderContainingReferenceNode as tpRef ->
+                            let containsTp = Microsoft.FSharp.Compiler.Driver.ContainsTypeProvider(tpRef.AssemblyPath)
+                            UIThread.RunSync (fun _ -> tpRef.ContainsTypeProvider <- containsTp)
+                        | _ -> ()
+                    sw.Stop()
+                    File.AppendAllText(timingFile, sprintf "%f %s\r\n" sw.Elapsed.TotalSeconds x.ProjectFile)
```

Comment:
The timing file stuff would obviously be removed before merging.  The feature could still be made optional via normal VS options, though...

On my laptop I saw that processing all references for a project is pretty fast - 0.29s was the longest it ever took for any particular project when opening `deployment.sln` cold.  The average time spent per project was around 0.05s.